Add ability to parse dicts of dicts in JSON to pass kwargs.

garymacindoe opened the pull request:
I have test functions with a large number of arguments:

    def test_function(arg0, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9):
        pass

Using the current version of ddt I can write my test cases as a dict of lists in JSON and get reasonably predictable automatically generated test names:

    {
        "test_case1": [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ],
        "test_case2": [ 0, 9, 8, 7, 6, 5, 4, 3, 2, 1 ]
    }

Due to the large number of arguments involved it rapidly becomes confusing mapping each argument value in the list to each named argument in the test function and it also makes refactoring the test function (specifically reordering the arguments) near impossible.

I have added a few lines to ddt which check whether the test arguments are a dict and passes them to the test function as kwargs rather than assuming they will always be a list and passing them as args.  This allows me to specify my test cases as a dict of dicts in JSON with named arguments while retaining predictable test case names and allowing the reordering of arguments in the test function:

    {
        "test_case1": { "arg0": 0,
                                "arg1": 1,
                                "arg2": 2,
                                "arg3": 3,
                                "arg4": 4,
                                "arg5": 5,
                                "arg6": 6,
                                "arg7": 7,
                                "arg8": 8,
                                "arg9": 9 },
        "test_case2": { "arg0": 0,
                                "arg1": 9,
                                "arg2": 8,
                                "arg3": 7,
                                "arg4": 6,
                                "arg5": 5,
                                "arg6": 4,
                                "arg7": 3,
                                "arg8": 2,
                                "arg9": 1 }
    }

txels commented:
Hey sorry this had fallen off my radar completely. Apologies for that. I will look at it shortly.

txels commented on ddt.py:
```diff
@@ -148,7 +148,10 @@ def _raise_ve(*args):
+                if isinstance(value, dict):
+                    add_test(test_name, func, **value)
+                else:
+                    add_test(test_name, func, value)
```
Looks simple and useful, although it may break backwards compatibility.

Can you illustrate with an example/test case in `test_example.py`?

txels commented:
Thanks for your contribution, and apologies again for the very late reply.

garymacindoe commented:
I've pulled in the changes from your master branch and updated my modifications to fit in with your refactored code.  Thanks for taking the time to review the changes.

I've also added a test case using a dict-of-dicts specified in a JSON file to specify kwargs to a test function.  The rest of the test cases continue to pass.

keradus commented:
Any news?
It would be really great to have that feature!

txels commented:
Thanks.
Aplogies again for losing awareness of this.

keradus commented:
Great, thanks !

clintonb commented:
Super-late to the party. I have an older library using version 1.0.0 that I want to update. This change broke the contract for `@file_data` defined in http://ddt.readthedocs.io/en/latest/example.html. The documentation should be updated to reflect the fact that dictionaries are expanded instead of passed as a single value (e.g. `**kwargs` instead of `kwargs`).
